File: classic puzzle-medium/numeral-system.rb
#https://www.codingame.com/training/medium/numeral-system
#NUMERAL SYSTEM
#concept: Parsing / Radix
ss=gets
s=ss.scan(/[0-9A-Z]+/)
r=2
ss.chars.map{
    t=_1=~/[A-Z]/?_1.ord-64:_1.to_i
    r=[t,r].max
}
radix=36
(36..2).step(-1).map{|k|
    a=s.map{_1.downcase.to_i(k)}
    if a[0]+a[1]==a[2]&&k>=r
        radix=[k,radix].min
    end
}
puts radix

File: classic puzzle-medium/numeral-system.py
#https://www.codingame.com/training/medium/numeral-system
#NUMERAL SYSTEM
#concept: Parsing / Radix
import re
import math

eq=input()
s=re.findall(r'[0-9A-Z]+',eq)
r=1
for x in eq:
    t=-1
    if x.isalpha(): t=ord(x)-65+10
    elif x.isdigit(): t=int(x)
    r=max(r,t)
ans=36
for radix in range(36,r,-1):
    a=[int(x,radix) for x in s]
    if sum(a[0:2])==a[2]:
        ans=min(radix,ans)
print(ans)
